src/GafferScene/Constraint.cpp

```diff
@@ -111,22 +147,358 @@ bool ptInPolygon( const Indexer indexer, const Imath::Vec2< T >& v, const int n
+
+			if( ( An <= 0.f ) && ( vi != vn ) )
+			{
+				const float l2 = ( vn - vi ).length2();
+				const float t = ( l2 > g_threshold )
+					? ( ( uv - vn ).length() / std::sqrt( l2 ) )
+					: 0.f;
+
+				Imath::V3f pv( 0.f );
+				Imath::V3f pn( 0.f );
+
+				float pvw = 0.f;
+				float pnw = 0.f;
+
+				for( int j = 0; j < n; ++j )
+				{
+					if( uvs( j ) == vi )
+					{
+						pv += points( j );
+						pvw += 1.f;
+					}
+					if( uvs( j ) == vn )
+					{
+						pn += points( j );
+						pnw += 1.f;
+					}
+				}
+
+				assert( pvw != 0.f );
+				assert( pnw != 0.f );
+
+				return
+					( pv / pvw ) * (       t ) +
+					( pn / pnw ) * ( 1.0 - t );
+			}
+
+			// NOTE : At this point Ad should only be zero when all three vertices and the uv are coincident
+
+			float const Ad = Ap * An;
+
+			if( Ad != 0.f )
+			{
+				const float w = Av / Ad;
+				p += points( i ) * w;
+				ws += w;
+			}
+		}
+	}
+
+	if( ws != 0.f )
+	{
+		p /= ws;
+	}
+
+	return p;
+}
+
+template< typename PIndexer, typename UVIndexer >
+Imath::V3f interpolateNonConvexPolygon( const PIndexer points, const UVIndexer uvs, const Imath::V2f& uv, const int n )
+{
+	// NOTE : a simple triangle fan from any vertex will cover the non convex polygon
+	//        the generated triangles may cover areas outside the polygon this is fine
+
+	Imath::V3f p( 0.f );
+
+	for( int i = 2; i < n; ++i )
+	{
+		const Imath::V2f tuv[ 3 ] =
+		{
+			uvs( 0 ),
+			uvs( i - 1 ),
+			uvs( i )
+		};
+
+		if( ptInPolygon( [ & tuv ]( const int ii ){ return tuv[ ii ]; }, uv, 3 ) )
+		{
+			const float w = ( tuv[ 2 ] - tuv[ 0 ] ).cross( tuv[ 1 ] - tuv[ 0 ] );
+
+			if( w == 0.f )
+			{
+				// NOTE : uv triangle numerically has zero area so is effectively a line or point.
+				//        position in 3d space is ambiguous unless the triangle in 3d space is a point.
+
+				const Imath::V3f& vp = points( i - 1 );
+
+				if(
+					( ( points( 0 ) - vp ).length2() <= g_threshold ) &&
+					( ( points( i ) - vp ).length2() <= g_threshold ) )
+				{
+					p = vp;
+
+					break;
+				}
+
+				throw IECore::InvalidArgumentException( (
+					boost::format(
+						"Gaffer::Constraint : UV coordinates \"%s\" map to ambiguous point(s) in 3d space."
+					) % ( uv ) ).str() );
+			}
+
+			// NOTE : ensure that the positive barycentric coordinates sum to a maximum of one.
+
+			const float b0 = std::max( 0.f, ( tuv[ 1 ] - tuv[ 2 ] ).cross( uv - tuv[ 1 ] ) / w );
+			const float b1 = std::max( 0.f, ( tuv[ 2 ] - tuv[ 0 ] ).cross( uv - tuv[ 2 ] ) / w );
+			const float b2 = std::max( 0.f, ( tuv[ 0 ] - tuv[ 1 ] ).cross( uv - tuv[ 0 ] ) / w );
+
+			const float bs = b0 + b1 + b2;
+
+			if( bs != 0.f )
+			{
+				p =
+					points(     0 ) * ( b0 / bs ) +
+					points( i - 1 ) * ( b1 / bs ) +
+					points( i     ) * ( b2 / bs );
+
+				break;
+			}
+		}
+	}
+
+	return p;
+}
+
+void constructLocalFrame( Imath::M44f& m, const Imath::V3f& p, const Imath::V3f& t, const Imath::V3f& b, const Imath::V3f& n )
```

Comment:
I'm wondering if this code would be simpler if it looked something like:
```
M44f constructFrameMatrix( V3f p, V3f x, V3f y, V3f z )
{
    M44f m;
      // Fill matrix ...
    return m;
}

void constructLocalFrame( Imath::M44f& m, const Imath::V3f& p, const Imath::V3f& t, const Imath::V3f& b, const Imath::V3f& n )
{
    if( t is non-zero && b is non-zero && t % b != 0 )
    {
        V3f y = b % t;
        if( nn.dot( y ) < 0.f )
        {
            y *= -1.f;
        }
        m = constructFrameMatrix( p, t.normalized(), y.normalized(), ( t % y ).normalized() );
    }
    else if( n is non-zero && t is non-zero && t % n != 0 )
    {
	V3f z = t % n;
        m = constructFrameMatrix( p, t.normalized(), ( z % t ).normalized(), z.normalized() );
    }
    else if( n is non-zero && b is non-zero && n % b != 0 )
    {
        V3f x = n % b;
        m = constructFrameMatrix( p, x.normalized(), ( b % x ).normalized(), b.normalized() );
    }
    else
    {
        m.setTranslation( p );
    }
}
```
I think this would achieve the same goal, and feels easier to understand to me?  Actually we could move all the normalizes into constructFrameMatrix to avoid repeating normalized()?

Reply:
I've refactored this code and added a test case that checks the frame at the poles of a sphere where one of the tangents becomes colinear with the normal. This test case fails when using the cross product to check for colinearity so the code uses the dot product.